ddl/job_table.go

```diff
@@ -0,0 +1,612 @@
+	defer ticker.Stop()
+	for {
+		if isChanClosed(d.ctx.Done()) {
+			return
+		}
+		if !variable.EnableConcurrentDDL.Load() || !d.isOwner() {
+			d.once.Store(true)
+			time.Sleep(time.Second)
+			continue
+		}
+		select {
+		case <-d.ddlJobCh:
+		case <-ticker.C:
+		case _, ok := <-notifyDDLJobByEtcdChGeneral:
+			if !ok {
+				logutil.BgLogger().Warn("[ddl] start general worker watch channel closed", zap.String("watch key", addingDDLJobGeneral))
+				notifyDDLJobByEtcdChGeneral = d.etcdCli.Watch(context.Background(), addingDDLJobGeneral)
+				time.Sleep(time.Second)
+				continue
+			}
+		case _, ok := <-notifyDDLJobByEtcdChReorg:
+			if !ok {
+				logutil.BgLogger().Warn("[ddl] start reorg worker watch channel closed", zap.String("watch key", addingDDLJobReorg))
+				notifyDDLJobByEtcdChReorg = d.etcdCli.Watch(context.Background(), addingDDLJobReorg)
+				time.Sleep(time.Second)
+				continue
+			}
+		case <-d.ctx.Done():
+			return
+		}
+		d.getDDLJobAndRun(sess, d.generalDDLWorkerPool, d.getGeneralJob)
+		d.getDDLJobAndRun(sess, d.reorgWorkerPool, d.getReorgJob)
+	}
+}
+
+func (d *ddl) getDDLJobAndRun(sess *session, pool *workerPool, getJob func(*session) (*model.Job, error)) {
+	wk, err := pool.get()
+	if err != nil || wk == nil {
+		logutil.BgLogger().Debug(fmt.Sprintf("[ddl] no %v worker available now", pool.tp()), zap.Error(err))
+		return
+	}
+
+	d.mu.RLock()
+	d.mu.hook.OnGetJobBefore(pool.tp().String())
+	d.mu.RUnlock()
+
+	job, err := getJob(sess)
+	if job == nil || err != nil {
+		if err != nil {
+			logutil.BgLogger().Warn("[ddl] get job met error", zap.Error(err))
+		}
+		pool.put(wk)
+		return
+	}
+	d.mu.RLock()
+	d.mu.hook.OnGetJobAfter(pool.tp().String(), job)
+	d.mu.RUnlock()
+	if err := d.markJobProcessing(sess, job); err != nil {
+		logutil.BgLogger().Info("[ddl] handle ddl job failed: mark job is processing meet error", zap.Error(err), zap.String("job", job.String()))
+		pool.put(wk)
+		return
+	}
+	d.delivery2worker(wk, pool, job)
+}
+
+func (d *ddl) delivery2worker(wk *worker, pool *workerPool, job *model.Job) {
+	injectFailPointForGetJob(job)
+	d.insertRunningDDLJobMap(job.ID)
+	d.wg.Run(func() {
+		metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Inc()
+		defer func() {
+			pool.put(wk)
+			d.deleteRunningDDLJobMap(job.ID)
+			asyncNotify(d.ddlJobCh)
+			metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Dec()
+		}()
+		// we should wait 2 * d.lease time to guarantee all TiDB server have finished the schema change.
+		// see waitSchemaSynced for more details.
+		if !d.isSynced(job) || d.once.Load() {
+			wk.waitSchemaSynced(d.ddlCtx, job, 2*d.lease)
+			d.once.Store(false)
+		}
+		if err := wk.HandleDDLJobTable(d.ddlCtx, job); err != nil {
+			logutil.BgLogger().Info("[ddl] handle ddl job failed", zap.Error(err), zap.String("job", job.String()))
+		}
+	})
+}
+
+func (d *ddl) markJobProcessing(sess *session, job *model.Job) error {
+	if !job.NotStarted() {
+		return nil
+	}
```

Comment:
This means the job is not processing until the second round of `runDDLJob` coming. Is it expected?

Reply:
no, it is **not** `NotStarted`, which will be marked processing only at the first round